transformations/transformations/data_offload.py

```diff
@@ -176,3 +181,335 @@ def remove_openmp_pragmas(self, routine, targets):
+            update_host += as_tuple(Pragma(keyword='acc', content='update self(' + ','.join(_acc_copyout) + ')'),)
+
+        # replace Loki pragmas with acc data/update pragmas
+        pragma_map = {}
+        for pragma in FindNodes(Pragma).visit(routine.body):
+            if pragma.keyword == 'loki':
+                if 'update_device' in pragma.content:
+                    if update_device:
+                        pragma_map[pragma] = update_device
+                    else:
+                        pragma_map[pragma] = None
+
+                if 'update_host' in pragma.content:
+                    if update_host:
+                        pragma_map[pragma] = update_host
+                    else:
+                        pragma_map[pragma] = None
+
+        routine.body = Transformer(pragma_map).visit(routine.body)
+
+        # build set of symbols to be offloaded
+        _var_set = reduce(operator.or_, [s.trafo_data[self._key]['var_set']
+                          for s in successors], set())
+        #build map of module imports corresponding to offloaded symbols
+        _modules = {}
+        _modules.update({k: v
+                         for s in successors if isinstance(s, SubroutineItem)
+                         for k, v in s.trafo_data[self._key]['modules'].items()})
+
+        # build new imports to add offloaded global vars to driver symbol table
+        new_import_map = {}
+        for s in _var_set:
+            if s in routine.symbol_map:
+                continue
+
+            if new_import_map.get(_modules[s], None):
+                new_import_map[_modules[s]] += as_tuple(s)
+            else:
+                new_import_map.update({_modules[s]: as_tuple(s)})
+
+        new_imports = ()
+        for k, v in new_import_map.items():
+            new_imports += as_tuple(Import(k, symbols=tuple(Variable(name=s, scope=routine) for s in v)))
+
+        # add new imports to driver subroutine sepcification
+        import_pos = 0
+        if (old_imports := FindNodes(Import).visit(routine.spec)):
+            import_pos = routine.spec.body.index(old_imports[-1]) + 1
+        if new_imports:
+            routine.spec.insert(import_pos, Comment(text=
+               '![Loki::GlobalVarOffload].....Adding global variables to driver symbol table for offload instructions'))
+            import_pos += 1
+            routine.spec.insert(import_pos, new_imports)
+
+    def process_kernel(self, routine, successors, item, targets):
+        """
+        Collect the set of module variables to be offloaded.
+        """
+
+        # filter out procedure imports
+        calls = [c.name for c in FindNodes(CallStatement).visit(routine.body)]
+        calls += FindInlineCalls().visit(routine.body)
+
+        imports = [i for i in FindNodes(Import).visit(routine.spec) if i.symbols not in calls]
+
+        # build map of modules corresponding to imports
+        import_mod = {s.name.lower(): i.module for i in imports for s in i.symbols}
```

Comment:
Considering the above comments, I think this could be built more easily like this:

```python
import_mod = CaseInsensitiveDict(
    (s.name, imprt) for imprt in routine.imports for s in imprt.symbols
    if s not in calls
)
```

Reply:
I prefer the use of `CaseInsensitiveDict`, thanks for the tip!